Added frames processing to AnalysisBase

rsexton2 opened the pull request:
Partially addresses #1985

Changes made in this Pull Request:
 - Make analysis base select arbitrary frames given by an array of integers (i.e., an array of frame indices) or booleans.
 
TODO (possibly), see #1985 
- accept FrameIterators


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

rsexton2 committed:
Added frames processing to AnalysisBase

- Fix #1985
- Added changes to AUTHORS CHANGELOG
- Added tests to test_base.py and changes to base.py

pep8speaks commented:
Hello @rsexton2! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-05-26 17:12:57 UTC

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,25 +295,31 @@ def run(self, start=None, stop=None, step=None, verbose=None):
+       
+
+        .. versionchanged:: 2.1.0
+            Added frames processing
```
I'd make this a bit more explicit "Added ability to iterate through trajectory by passing a list of frame indices".

rsexton2 replied:
I changed the description to be more explicit, thank you.

IAlibay replied:
I would preferentially add the more explicit versionchanged text here rather than under `_setup_frames` since `run` is exposed in the docs but `_setup_frames` isn't.

rsexton2 replied:
I added the versionchanged text to run

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -270,7 +283,8 @@ def _conclude(self):
         pass  # pylint: disable=unnecessary-pass
 
-    def run(self, start=None, stop=None, step=None, verbose=None):
+    def run(self, start=None, stop=None, step=None, verbose=None,
```
nitpicking, but I'd place frames before verbose here, mostly a visual thing but it looks nicer to have start/stop/step/frame bunched up together

orbeckst replied:
(I think that was me suggesting to keep the diff minimal... but you're absolutely right that frames should be with the other frame selections)

rsexton2 replied:
I switched these around

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -238,17 +239,29 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None):
         """
         self._trajectory = trajectory
-        start, stop, step = trajectory.check_slice_indices(start, stop, step)
+        if frames is not None:
```
Appropriate tests will be necessary here.

rsexton2 replied:
I added tests to ensure frames isn't used with start, stop or step

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -238,17 +239,29 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None):
+        else:
+            start, stop, step = trajectory.check_slice_indices(start, stop, step)
+            slicer = slice(start, stop, step)
+        self._sliced_trajectory = trajectory[slicer]
```
I _think_ using a frameiterator changes the behaviour of the iteration.

So if I do:

```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import TPR, XTC

u = mda.Universe(TPR, XTC)

iterator = u.trajectory[[0,1,2]]

for ts in iterator:
    pass

print(u.trajectory.ts)
```
Ouput: < Timestep 2 with unit cell dimensions [79.98359 79.98359 79.98358 60.      60.      90.     ] >

Whilst:

```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import TPR, XTC

u = mda.Universe(TPR, XTC)

for ts in u.trajectory[:3]
    pass

print(u.trajectory.ts)
```
Ouput: < Timestep 0 with unit cell dimensions [80.017006 80.017006 80.017006 60.       60.       90.      ] >

This is a general thing with the way we iterate through a trajectory based on an index selection (not limited to this code here), and we should probably fix this? But I'd say it's probably a bad idea for the two approaches differ like this for AnalysisBase.

IAlibay replied:
I've opened a separate issue, I'm not actually sure if it's intended behaviour.

orbeckst replied:
Thanks for opening  #3416.

As a band-aid solution we could add a `self._trajectory.rewind()` after the `for i, ts in enumerate(self._sliced_trajectory)` loop. This would ensure consistent behavior of all AnalysisBase classes. It would just need a comment that references issue #3416 so that we can remove it again if that gets fixed/changed.

IAlibay replied:
Sure I'm fine with this as a temporary measure. We should implement the fix in this PR so that we don't forget.

rsexton2 replied:
I added the line to rewind with a comment referencing the issue

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -238,17 +239,29 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None):
         self.step = step
-        self.n_frames = len(range(start, stop, step))
+        self.n_frames = len(self._sliced_trajectory)
         self.frames = np.zeros(self.n_frames, dtype=int)
```
I can see this becoming confusing real quickly, we're offering `frames` as an input argument to run and then offering a different thing contained in the `frames` user-facing variable. Thoughts?

Maybe we should call the input argument `indices` instead?

orbeckst replied:
I see your point. Although in principle `self.frames` should be equivalent to what the user inputs in `frames` in the sense that running a second time `a.run(frames=a.frames)` should give the identical answer (assuming that `ts.frame` is the true frame index), which I like. If the input `frames` was a boolean array or a `slice()` then `self.frames` will be an int array of frame indices but _should_ be functionally equivalent.

I am not too fond of calling the input `frames` kwarg `indices` because (I think) we typically use it for indexing atoms. (?)

orbeckst replied:
@IAlibay can you be convinced to leave the input `frames` as originally proposed or do you have strong opinions on the matter?

IAlibay replied:
It's a little bit messy but I'm not too fussed.

I'm not sure what (if anything) would need adding, but do we want to alter the attributes description for frame to make this clearer?

rsexton2 replied:
@IAlibay How should I alter the attribute description for `frames`? 
It is currently "array of Timestep frame indices. Only exists after calling AnalysisBase.run"
Something like "array of Timestep frame indices. Only exists after calling AnalysisBase.run. If the 'frames' 
option is used in .run(), the input for 'frames' will be returned"

IAlibay replied:
With hindsight, let's leave the attribute as it is. I'm not really sure how to amend (or if it's at all possible) it without making it more complex.

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,25 +295,31 @@ def run(self, start=None, stop=None, step=None, verbose=None):
         step : int, optional
             number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory
```
We'll need some documentation / examples for this.

rsexton2 replied:
Added an example here with two uses of 'frames'

IAlibay reviewed (changes requested):
Couple of initial comments.

Main thing here is that I'm a little bit apprehensive about the differing behaviour about which frame you end up with at the end of for the frame iteration depending on if you use start/stop/step or a list of indices.

orbeckst commented:
@rsexton2 please ping me when you're ready (remove draft status) for review and I am happy to have a look; while you're hacking away at tests, please already address @IAlibay's comments, which are all very good & relevant. Also have a look at what PEP8speaks complains about, please.

rsexton2 commented:
@orbeckst I added tests but still need to test slicing by a boolean array.
I think I addressed most of @IAlibay comments in the changes and fixed pep8 things

orbeckst commented:
For some reason the CI was waiting for approval — sorry @rsexton2 , I didn't notice.

Given that you have added tests, do you want to set the status to "ready to review", i.e., remove the draft status?

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -197,6 +197,45 @@ def test_start_stop_step(u, run_kwargs, frames):
+    ({'frames': [4, 5, 6, 7, 8, 9]}, np.arange(4, 10)),
+    ({'frames': [0, 2, 4, 6, 8]}, np.arange(0, 10, 2)),
+    ({'frames': [4, 6, 8]}, np.arange(4, 10, 2)),
+    ({'frames': [True, True, False, True, False, True, True, False, True,
```
Also add `[0, 3, 4, 3, 5]` with a repeated frame.

Otherwise I like that you're checking `an.frames` !

rsexton2 replied:
Is the proper behavior that it will run frame 3 twice?

orbeckst replied:
yes

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -197,6 +197,45 @@ def test_start_stop_step(u, run_kwargs, frames):
+
+
+@pytest.mark.parametrize('run_kwargs, frames', [
+    ({'start': 4, 'frames': [4, 5, 6, 7, 8, 9]}, np.arange(4, 10)),
```
What happens when you have `start=0, stop=0, step=0` ?

rsexton2 replied:
This slipped through.
I changed _setup_frames to fail if start/stop/step aren't all None

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -197,6 +197,45 @@ def test_start_stop_step(u, run_kwargs, frames):
+    assert_equal(an.frames, frames, err_msg=FRAMES_ERR)
+
+
+@pytest.mark.parametrize('run_kwargs, frames', [
```
remove the frames part from the parametrization. You never check against it, so this is just noise and confuses.

rsexton2 replied:
done

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -197,6 +197,45 @@ def test_start_stop_step(u, run_kwargs, frames):
+    ({'stop': 6, 'step': 2, 'frames': [0, 2, 4, 6]}, np.arange(0, 7, 2)),
+    ({'start': 4, 'step': 2, 'frames': [4, 6, 8]}, np.arange(4, 10, 2)),
+])
+def test_frame_fail(u, run_kwargs, frames):
```
Remove the reference value that is never tested
```suggestion
def test_frame_fail(u, run_kwargs):
```

rsexton2 replied:
done

orbeckst reviewed (changes requested):
This looks already very good and your test coverage is great!

I only found minor things — I'll approve once you addressed my comments.

The only major remaining issue is some example docs as requested by @IAlibay and I will let him review that aspect (i.e., all reviewers must approve).

IAlibay commented on .github/pr_path_labeler.yml:
```diff
@@ -0,0 +1,26 @@
+# Labeler triggered by .github/workflows/labels.yml
```
Strange, why are these files here? Did the merge not go right?

orbeckst replied:
I'm not sure why they showed up. _Should_ be harmless when squash-merged, right?

rsexton2 replied:
I rebased against an updated develop and squashed everything into 1 commit

IAlibay commented on package/CHANGELOG:
```diff
@@ -13,13 +13,14 @@ The rules for this file:
 Fixes
 
 Enhancements
+  * Added frames processing to AnalysisBase.run (Issue #1985)
```
Can we use something similar to the versionchanged entry here? Whilst I'm sure it's not something most people read, it is where we direct folks to read about new changes, so being a bit more descriptive doesn't hurt.

rsexton2 replied:
Added more explicit description

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_base.py:
```diff
@@ -197,6 +197,47 @@ def test_start_stop_step(u, run_kwargs, frames):
     assert_almost_equal(an.times, frames+1, decimal=4, err_msg=TIMES_ERR)
 
 
```
It'd be too much to ask to add a test for every analysis method, but could we maybe throw in an extra test for RMSD that shows that passing frames works with that as expected?

rsexton2 replied:
I added a test to test_rmsd.py

IAlibay reviewed (changes requested):
Sorry only had time for a quick glance at it. Do ping me when you have the documentation written.

orbeckst commented:
@IAlibay I assigned the PR to you as you're the last one with outstanding change requests.

I don't know why some of the commits that are already in master show up here, perhaps they came from pushing the "update with develop" button (or whatever it's called when GitHub tells you a branch is out of date). I _think_ it will be fine when squash-merging but if there are any concerns then possibly rebasing this branch against and updated develop??

rsexton2 committed:
Added rmsd test and 'frames' examples in docs

rsexton2 committed:
fixed pep8

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,26 +298,51 @@ def run(self, start=None, stop=None, step=None, verbose=None):
-            # logger.info("--> Doing frame {} of {}".format(i+1, self.n_frames))
             self._single_frame()
+        # rewind added for behavioral consistency, Issue #3416
+        self._trajectory.rewind()
```
Thanks, could you just add a quick test to check that this is working as intended for now? (nothing fancy, just a quick check that the trajectory rewinds back to 0 will do)

rsexton2 replied:
done

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,26 +298,51 @@ def run(self, start=None, stop=None, step=None, verbose=None):
             Turn on verbosity
+
+
+    Example
+    -------
```
Should this be indented by two the right?

rsexton2 replied:
You're right, I fixed it

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,26 +298,51 @@ def run(self, start=None, stop=None, step=None, verbose=None):
+       u = mda.Universe(TPR, XTC)
+
+       rmsd = rms.RMSD(u, select='name CA')
+       # A list of frames to run the analysis on can be passed
```
Could you also throw in an example using start/stop/step? That way we have every use case here?

rsexton2 replied:
done

IAlibay reviewed (changes requested):
Thanks, just a couple of hopefully final changes.

rsexton2 committed:
Added rewind test, fixed .run example

rsexton2 committed:
Fixed pep8

IAlibay reviewed (approved):
lgtm assuming CI returns green

rsexton2 commented:
git bisect says commit c35458b2b004a34a595c1c881be826cb9e0df460 is the one that makes the test testsuite/MDAnalysisTests/analysis/test_rdf.py::test_double_run FAILED

Locally testsuite/MDAnalysisTests/analysis/test_rdf.py::test_count FAILED
but not in the CI

```
c35458b2b004a34a595c1c881be826cb9e0df460 is the first bad commit
commit c35458b2b004a34a595c1c881be826cb9e0df460
Author: Ricky Sexton <rsexton2@asu.edu>
Date:   Fri Sep 10 15:49:19 2021 -0700

    Added rmsd test and 'frames' examples in docs

 package/CHANGELOG                              |  3 ++-
 package/MDAnalysis/analysis/base.py            | 18 ++++++++++++++++++
 testsuite/MDAnalysisTests/analysis/test_rms.py |  7 +++++++
 3 files changed, 27 insertions(+), 1 deletion(-)
```

The only thing that is added here is the rewind.
I'm not sure how this is affecting InterRDF

rsexton2 commented:
@richardjgowers @VOD555 
Do you have any idea what is happening with InterRDF and trajectory.rewind() ?

IAlibay commented:
Just to follow up from discord - the issue seems to be from SingleFrameBase implementing an `_iter_` that doesn't rewind, which means that FrameIteratorAll rewinds on multi-frame trajectories but not on single frame readers.

In this particularly case - prior to this PR the `sels` fixture in test_rdf changes the coordinates and then keeps them that way. Whilst now the rewind specifically resets the timestep coordinates, hence yielding the test failure.

orbeckst commented:
What do we need to do so that this PR can move forward?

orbeckst commented:
As outlined in https://github.com/MDAnalysis/mdanalysis/pull/3415#issuecomment-922109775, iteration for single frame "trajectories" behaves differently from the way our multi-frame iterators _ought to_ behave (see also #3416 ). 

Could someone with proper understanding of the problem (@richardjgowers or @IAlibay ?) please raise an issue with a conclusion of what we will do about it? (I don't think it belongs into #3416, does it?) Then we can fix that problem and hopefully this PR can then be merged without additional changes.

IAlibay commented:
> What do we need to do so that this PR can move forward?

Assuming we want to change the behaviour of singleframe, a quick fix here would be to just change the failing test to not use a function scoped fixture and instead just do the frame manipulation twice. Either way we'll need to do that once we implement the singleframe change/fix.

I'll raise an issue.

orbeckst committed:
Merge branch 'develop' into frames_analysis

orbeckst committed:
Merge branch 'develop' into frames_analysis

jbarnoud commented:
I only had a very brief look, but since this PR already slices the trajectory, what makes it challenging to accept an already sliced trajectory as well?

If this PR goes through without accepting a sliced trajectory, the description of the PR should be edited so it does not close the original issue.

orbeckst commented:
In the original discussion following https://github.com/MDAnalysis/mdanalysis/issues/1985#issuecomment-913865103 we had decided to start out conservatively with only explicit start/stop/step. I agree that being able to hand over FrameIterators would make it really complete. 

It makes sense exploring expanding this PR once we get the tests to pass, which requires #3642 IIRC.

orbeckst commented:
For right now I changed the issue description to reflect the code that is present.

orbeckst commented:
PR #3642 needs to get merged before we can finalize this PR.

IAlibay commented on package/AUTHORS:
```diff
@@ -171,6 +171,7 @@ Chronological list of authors
   - Marcelo C. R. Melo
   - Mark D. Driver
   - Kevin Boyd
+  
```
```suggestion
```
We don't usually have an empty line between year

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -238,17 +239,32 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None):
         .. versionchanged:: 1.0.0
             Added .frames and .times arrays as attributes
+
+        .. versionchanged:: 2.1.0
```
```suggestion
        .. versionchanged:: 2.2.0
```

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,26 +298,54 @@ def run(self, start=None, stop=None, step=None, verbose=None):
             self.times[i] = ts.time
-            # logger.info("--> Doing frame {} of {}".format(i+1, self.n_frames))
             self._single_frame()
+        # rewind added for behavioral consistency, Issue #3416
+        self._trajectory.rewind()
```
```suggestion
```
_should_ be fixed now

IAlibay reviewed (commented):
I think just these and this is ready to go (going to self-merge to speed things up).

IAlibay committed:
Apply suggestions from code review

IAlibay committed:
Merge branch 'develop' into frames_analysis

IAlibay commented:
Assuming CI returns green I'm happy for this to get merged - @orbeckst do you want to have a final look?

IAlibay committed:
Add codeblock to example

orbeckst commented:
Please go ahead at your leisure — don’t have the bandwidth and don’t want to hold up things. Thank you for looking after this one and the related PR!!

> Am 5/13/22 um 16:36 schrieb Irfan Alibay ***@***.***>:
> 
> ﻿
> Assuming CI returns green I'm happy for this to get merged - @orbeckst do you want to have a final look?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

lilyminium reviewed (approved):
LGTM, thanks @rsexton2, @IAlibay, @orbeckst.

orbeckst committed:
Merge branch 'develop' into frames_analysis

orbeckst commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -281,25 +298,54 @@ def run(self, start=None, stop=None, step=None, verbose=None):
+        -------
+
+        .. code-block:: python
+        
```
```suggestion

```

orbeckst committed:
Apply suggestions from code review

fix pep8 issue

orbeckst committed:
improved AnalysisBase documentation

- outline general ideas and features
- show example (RMSD) with different frame selections
- fixed generic example on top analysis page

IAlibay commented:
@orbeckst is this good to re-review or are you still working on the docs?

orbeckst commented:
Done with docs, good for most final of final reviews (I hope)—feel free to just edit as you see fit. 

The main thing was to move a specific example out of run() because it would be really confusing for all tools to reference RMSD in their inherited run(). 

> Am 5/21/22 um 05:27 schrieb Irfan Alibay ***@***.***>:
> 
> ﻿
> @orbeckst is this good to re-review or are you still working on the docs?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

orbeckst commented:
Can we merge this one ... finally??

IAlibay commented:
> Can we merge this one ... finally??

Sorry it's on my task list, I'll have a final look in the morning

IAlibay commented on package/MDAnalysis/analysis/base.py:
```diff
@@ -20,11 +20,105 @@
+
+The :class:`Results` and :class:`AnalysisBase` classes are the essential
+building blocks for almost all MDAnalysis tools in the
+:mod:`MDAnalysis.analysis` module. They are supposed to be easily useable and
```
```suggestion
:mod:`MDAnalysis.analysis` module. They aim to be easily useable and
```

IAlibay reviewed (approved):
lgtm just the one word change which I'll commit

IAlibay committed:
Update package/MDAnalysis/analysis/base.py

IAlibay commented:
Thanks @rsexton2 ! Sorry it took so long to merge, but it fixed some pretty big bugs in MDAnalysis along the way! 🎉
